components/camel-kafka/src/main/java/org/apache/camel/component/kafka/KafkaProducer.java

```diff
@@ -143,167 +145,260 @@ protected void doStop() throws Exception {
-        return Collections.singletonList(new KeyValueHolder<Object, ProducerRecord>((Object) exchange, record)).iterator();
+    protected KeyValueHolder<Object, ProducerRecord<Object, Object>> createKeyValueHolder(Exchange exchange, Message message) {
+        ProducerRecord<Object, Object> record = createRecord(exchange, message);
+
+        return new KeyValueHolder<>(exchange, record);
     }
 
-    private List<Header> getPropagatedHeaders(Exchange exchange, KafkaConfiguration getConfiguration) {
-        HeaderFilterStrategy headerFilterStrategy = getConfiguration.getHeaderFilterStrategy();
-        KafkaHeaderSerializer headerSerializer = getConfiguration.getHeaderSerializer();
-        return exchange.getIn().getHeaders().entrySet().stream()
-                .filter(entry -> shouldBeFiltered(entry, exchange, headerFilterStrategy))
-                .map(entry -> getRecordHeader(entry, headerSerializer)).filter(Objects::nonNull).collect(Collectors.toList());
+    private String evaluateTopic(Message message) {
+        String topic = configuration.getTopic();
+
+        // must remove header so it's not propagated
+        Object overrideTopic = message.removeHeader(KafkaConstants.OVERRIDE_TOPIC);
+        if (overrideTopic != null) {
+            LOG.debug("Using override topic: {}", overrideTopic);
+            topic = overrideTopic.toString();
+        }
+
+        if (topic == null) {
+            // if topic property was not received from configuration or header
+            // parameters take it from the remaining URI
+            topic = URISupport.extractRemainderPath(URI.create(endpoint.getEndpointUri()), true);
+        }
+        return topic;
     }
 
-    private boolean shouldBeFiltered(
-            Map.Entry<String, Object> entry, Exchange exchange, HeaderFilterStrategy headerFilterStrategy) {
-        return !headerFilterStrategy.applyFilterToCamelHeaders(entry.getKey(), entry.getValue(), exchange);
+    private boolean isIterable(Object body) {
+        if (body instanceof Iterable || body instanceof Iterator) {
+            return true;
+        }
+
+        return false;
     }
 
-    private RecordHeader getRecordHeader(Map.Entry<String, Object> entry, KafkaHeaderSerializer headerSerializer) {
-        byte[] headerValue = headerSerializer.serialize(entry.getKey(), entry.getValue());
-        if (headerValue == null) {
-            return null;
+    private Iterator<Object> getObjectIterator(Object body) {
+        Iterator<Object> iterator = null;
+        if (body instanceof Iterable) {
+            iterator = ((Iterable<Object>) body).iterator();
+        } else if (body instanceof Iterator) {
+            iterator = (Iterator<Object>) body;
         }
-        return new RecordHeader(entry.getKey(), headerValue);
+        return iterator;
+    }
+
+    private List<Header> getPropagatedHeaders(Exchange exchange, Message message) {
+        return message.getHeaders().entrySet().stream()
```

Comment:
This can be optimized to not use all this streaming for a basic task that a for loop would be faster

Reply:
+1. I had measured the streaming one and the for each loop one and there was hardly any change ... but that was with doing 2 passes instead of one. I'll change w/ a for each and try/measure.